File: platforms/ios/Dezel/Dezel/Layout/Layout.swift
import Foundation

/**
 * Contextual informations required by each layout node.
 * @class Layout
 * @since 0.4.0
 */
open class Layout {

	//--------------------------------------------------------------------------
	// MARK: Properties
	//--------------------------------------------------------------------------

	/**
	 * Whether the layout is invalid.
	 * @property invalid
	 * @since 0.4.0
	 */
	open var invalid: Bool {
		return DLLayoutIsInvalid(self.handle)
	}

	/**
	 * Whether the layout is resolving.
	 * @property resolving
	 * @since 0.4.0
	 */
	open var resolving: Bool {
		return DLLayoutIsResolving(self.handle)
	}

	/**
	 * The layout's root node.
	 * @property root
	 * @since 0.4.0
	 */
	open var root: LayoutNode? {
		willSet {
			DLLayoutSetRoot(self.handle, newValue?.handle)
		}
	}

	/**
	 * The layout's viewport width.
	 * @property viewportWidth
	 * @since 0.4.0
	 */
	open var viewportWidth: CGFloat = 0 {
		willSet {
			DLLayoutSetViewportWidth(self.handle, Double(newValue))
		}
	}

	/**
	 * The layout's viewport height.
	 * @property viewportHeight
	 * @since 0.4.0
	 */
	open var viewportHeight: CGFloat = 0 {
		willSet {
			DLLayoutSetViewportHeight(self.handle, Double(newValue))
		}
	}

	/**
	 * The layout's scale.
	 * @property scale
	 * @since 0.4.0
	 */
	open var scale: CGFloat = 1 {
		willSet {
			DLLayoutSetScale(self.handle, Double(newValue))
		}
	}

	/**
	 * @property handle
	 * @since 0.4.0
	 * @hidden
	 */
	private(set) public var handle: DLLayoutRef

	/**
	 * @property layoutBeganCallbacks
	 * @since 0.4.0
	 * @hidden
	 */
	private var layoutBeganCallbacks: [(() -> Void)] = []

	/**
	 * @property layoutEndedCallbacks
	 * @since 0.4.0
	 * @hidden
	 */
	private var layoutEndedCallbacks: [(() -> Void)] = []

	//--------------------------------------------------------------------------
	// MARK: Methods
	//--------------------------------------------------------------------------

	/**
	 * @constructor
	 * @since 0.4.0
	 */
	internal init() {
		self.handle = DLLayoutCreate()
		DLLayoutSetLayoutBeganCallback(self.handle, layoutBeganCallback)
		DLLayoutSetLayoutEndedCallback(self.handle, layoutEndedCallback)
		DLLayoutSetData(self.handle, UnsafeMutableRawPointer(value: self))
	}

	/**
	 * @destructor
	 * @since 0.4.0
	 */
	deinit {
		DLLayoutGetData(self.handle).release()
		DLLayoutDelete(self.handle)
	}

	/**
	 * Requests a callback that will be executed when the global layout begins.
	 * @method requestLayoutBeganCallback
	 * @since 0.4.0
	 */
	public func requestLayoutBeganCallback(_ callback: @escaping () -> Void) {
		self.layoutBeganCallbacks.append(callback)
	}

	/**
	 * Requests a callback that will be executed when the global layout finishes.
	 * @method requestLayoutEndedCallback
	 * @since 0.4.0
	 */
	public func requestLayoutEndedCallback(_ callback: @escaping () -> Void) {
		self.layoutEndedCallbacks.append(callback)
	}

	//--------------------------------------------------------------------------
	// MARK: Methods
	//--------------------------------------------------------------------------

	/**
	 * @method dispatchLayoutBeganEvent
	 * @since 0.4.0
	 * @hidden
	 */
	internal func dispatchLayoutBeganEvent() {
		self.layoutBeganCallbacks.forEach { $0() }
		self.layoutBeganCallbacks.removeAll()
	}

	/**
	 * @method dispatchLayoutEndedEvent
	 * @since 0.4.0
	 * @hidden
	 */
	internal func dispatchLayoutEndedEvent() {
		self.layoutEndedCallbacks.forEach { $0() }
		self.layoutEndedCallbacks.removeAll()
	}
}

/**
 * @const layoutBeganCallback
 * @since 0.4.0
 * @hidden
 */
private let layoutBeganCallback: @convention(c) (DLLayoutRef?) -> Void = { ptr in
	if let layout = DLLayoutGetData(ptr).value as? Layout {
		layout.dispatchLayoutBeganEvent()
	}
}

/**
 * @const layoutEndedCallback
 * @since 0.4.0
 * @hidden
 */
private let layoutEndedCallback: @convention(c) (DLLayoutRef?) -> Void = { ptr in
	if let layout = DLLayoutGetData(ptr).value as? Layout {
		layout.dispatchLayoutEndedEvent()
	}
}

File: platforms/android/Dezel/dezel/src/main/java/ca/logaritm/dezel/layout/Layout.kt
package ca.logaritm.dezel.layout

import ca.logaritm.dezel.extension.Delegates

/**
 * The layout manager.
 * @class Layout
 * @since 0.4.0
 */
open class Layout {

	//--------------------------------------------------------------------------
	// Properties
	//--------------------------------------------------------------------------

	/**
	 * @property invalid
	 * @since 0.4.0
	 */
	open var invalid: Boolean = false
		get() = LayoutExternal.isInvalid(this.handle)

	/**
	 * @property resolving
	 * @since 0.4.0
	 */
	open var resolving: Boolean = false
		get() = LayoutExternal.isResolving(this.handle)

	/**
	 * @property root
	 * @since 0.4.0
	 */
	open var root: LayoutNode? by Delegates.OnSetOptional<LayoutNode>(null) { value ->
		LayoutExternal.setRoot(this.handle, value?.handle ?: 0)
	}

	/**
	 * @property viewportWidth
	 * @since 0.4.0
	 */
	open var viewportWidth: Float by Delegates.OnSet(0f) { value ->
		LayoutExternal.setViewportWidth(this.handle, value.toDouble())
	}

	/**
	 * @property viewportHeight
	 * @since 0.4.0
	 */
	open var viewportHeight: Float by Delegates.OnSet(0f) { value ->
		LayoutExternal.setViewportHeight(this.handle, value.toDouble())
	}

	/**
	 * @property scale
	 * @since 0.4.0
	 */
	open var scale: Float by Delegates.OnSet(1f) { value ->
		LayoutExternal.setScale(this.handle, value.toDouble())
	}

	/**
	 * @property handle
	 * @since 0.4.0
	 * @hidden
	 */
	public var handle: Long
		private set

	/**
	 * @property layoutBeganCallbacks
	 * @since 0.4.0
	 * @hidden
	 */
	private var layoutBeganCallbacks: MutableList<LayoutBeganCallback> = mutableListOf()

	/**
	 * @property layoutEndedCallbacks
	 * @since 0.4.0
	 * @hidden
	 */
	private var layoutEndedCallbacks: MutableList<LayoutEndedCallback> = mutableListOf()

	//--------------------------------------------------------------------------
	// Methods
	//--------------------------------------------------------------------------

	/**
	 * @constructor
	 * @since 0.4.0
	 * @hidden
	 */
	init {
		this.handle = LayoutExternal.create(this)
		LayoutReference.register(this)
	}

	/**
	 * Requests a handler that will be executed when the global layout begins.
	 * @method requestLayoutBeganCallback
	 * @since 0.4.0
	 */
	public fun requestLayoutBeganCallback(callback: (() -> Unit)) {
		this.layoutBeganCallbacks.add(LayoutBeganCallback(callback))
	}

	/**
	 * Requests a handler that will be executed when the global layout finishes.
	 * @method requestLayoutEndedCallback
	 * @since 0.4.0
	 */
	open fun requestLayoutEndedCallback(callback: (() -> Unit)) {
		this.layoutEndedCallbacks.add(LayoutEndedCallback(callback))
	}

	/**
	 * @method dispatchLayoutBeganEvent
	 * @since 0.4.0
	 * @hidden
	 */
	@Suppress("unused")
	private fun dispatchLayoutBeganEvent() {
		this.layoutBeganCallbacks.forEach { it.callback() }
		this.layoutBeganCallbacks.clear()
	}

	/**
	 * @method dispatchLayoutEndedEvent
	 * @since 0.4.0
	 * @hidden
	 */
	@Suppress("unused")
	private fun dispatchLayoutEndedEvent() {
		this.layoutEndedCallbacks.forEach { it.callback() }
		this.layoutEndedCallbacks.clear()
	}

	//--------------------------------------------------------------------------
	// Classes
	//--------------------------------------------------------------------------

	/**
	 * @class LayoutBeganCallback
	 * @since 0.4.0
	 * @hidden
	 */
	private data class LayoutBeganCallback(val callback: (() -> Unit))

	/**
	 * @class LayoutEndedCallback
	 * @since 0.4.0
	 * @hidden
	 */
	private data class LayoutEndedCallback(val callback: (() -> Unit))
}